pkg/resource/resource.go

```diff
@@ -227,3 +229,52 @@ func IsBound(b Bindable) bool {
 func IsConditionTrue(c v1alpha1.Condition) bool {
 	return c.Status == corev1.ConditionTrue
 }
+
+// ApplyOptions configure how changes are applied to an object.
+type ApplyOptions struct {
+	// ControllersMustMatch requires any existing object to have a controller
+	// reference, and for that controller reference to match the controller
+	// reference of the supplied object.
+	ControllersMustMatch bool
+}
+
+// An ApplyOption configures how changes are applied to an object.
+type ApplyOption func(a *ApplyOptions)
+
+// ControllersMustMatch requires any existing object to have a controller
+// reference, and for that controller reference to match the controller
+// reference of the supplied object.
+func ControllersMustMatch() ApplyOption {
+	return func(a *ApplyOptions) {
+		a.ControllersMustMatch = true
+	}
+}
+
+// Apply changes to the supplied object. The object will be created if it does
+// not exist, or patched if it does.
+func Apply(ctx context.Context, c client.Client, o runtime.Object, ao ...ApplyOption) error {
+	opts := &ApplyOptions{}
+	for _, fn := range ao {
+		fn(opts)
+	}
+
+	m, ok := o.(metav1.Object)
+	if !ok {
+		return errors.New("cannot access object metadata")
+	}
+
+	existing := o.DeepCopyObject()
+	err := c.Get(ctx, types.NamespacedName{Name: m.GetName(), Namespace: m.GetNamespace()}, existing)
+	if kerrors.IsNotFound(err) {
+		return errors.Wrap(c.Create(ctx, o), "cannot create object")
+	}
+	if err != nil {
+		return errors.Wrap(err, "cannot get object")
+	}
+
+	if opts.ControllersMustMatch && !meta.HaveSameController(o.(metav1.Object), m) {
+		return errors.New("existing object has a different (or no) controller")
+	}
+
+	return errors.Wrap(c.Patch(ctx, existing, client.MergeFrom(o)), "cannot patch object")
```

Comment:
`client.MergeFrom` creates a merge patch JSON from the difference but that merge patch aims to make the final state of the object exactly like `o`. Meaning if any field is late-inited and you didn't work on the most up-to-date version, produced merge patch will have a `null` for that field and remove it. So, the ideal workflow for `MergeFrom` is that you fetch the object, make the changes and then issue a `MergeFrom` patch. However, here the object fetched and directly supplied for patch, so you'll delete the fields that are populated but you don't know about.

I've used [`ConstantPatch`](https://github.com/crossplane/templating-controller/blob/cb1327d/pkg/controllers/templating_reconciler.go#L182) in templating-controller after realizing that, which solved the problem of deleting the fields that `o` is not aware and are populated.

Reply:
Aha, thanks! I initially copied your approach but later switched to `MergeFrom` simply because it seemed more succinct. Perhaps this behaviour should be configurable using an `ApplyOption`? If so, I wonder whether the default should be to remove or preserve any unexpected fields. The former seems safer, but the latter is arguably less surprising.

Reply:
Just a note that `ConstantPatch` is deprecated in favor of `RawPatch`. Would likely make sense to go ahead and use `RawPatch` so that we do not have to fix in coming updates 👍 https://github.com/kubernetes-sigs/controller-runtime/blob/9f8aab6551a3ecc211b6c6078245899495b7e0f5/pkg/client/patch.go#L57

Reply:
@hasheddan I noticed that, but `RawPatch` doesn't seem to exist yet in the version of controller-runtime we're using.

Reply:
@negz do you think it would be worth bumping our version or no?

Reply:
> do you think it would be worth bumping our version or no?

I ended up not using `ConstantPatch`, but instead implementing the `Patch` interface in a fashion similar to `ConstantPatch`.